src/uu/wc/src/count_bytes.rs

```diff
@@ -0,0 +1,95 @@
+use super::{WcResult, WordCountable};
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+use std::fs::OpenOptions;
+use std::io::ErrorKind;
+
+#[cfg(unix)]
+use libc::S_IFREG;
+#[cfg(unix)]
+use nix::sys::stat::fstat;
+#[cfg(any(target_os = "linux", target_os = "android"))]
+use std::os::unix::io::{AsRawFd, RawFd};
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+use libc::S_IFIFO;
+#[cfg(any(target_os = "linux", target_os = "android"))]
+use nix::fcntl::{splice, SpliceFFlags};
+#[cfg(any(target_os = "linux", target_os = "android"))]
+use nix::unistd::pipe;
+
+const BUF_SIZE: usize = 16384;
+
+/// This is a Linux-specific function to count the number of bytes using the
+/// `splice` system call, which is faster than using `read`.
+#[inline]
+#[cfg(any(target_os = "linux", target_os = "android"))]
+fn count_bytes_using_splice(fd: RawFd) -> nix::Result<usize> {
+    let null_file = OpenOptions::new()
+        .write(true)
+        .open("/dev/null")
+        .map_err(|_| nix::Error::last())?;
+    let null = null_file.as_raw_fd();
+    let (pipe_rd, pipe_wr) = pipe()?;
+
+    let mut byte_count = 0;
+    loop {
+        let res = splice(fd, None, pipe_wr, None, BUF_SIZE, SpliceFFlags::empty())?;
+        if res == 0 {
+            break;
+        }
+        let res = splice(pipe_rd, None, null, None, BUF_SIZE, SpliceFFlags::empty())?;
+        byte_count += res;
```

Comment:
Do you happen to know if the second `splice()` could potentially write less than the first `splice()`?  If so, this should just add the `res` from the first one to `byte_count` (otherwise `byte_count` could be smaller than it should be).

Also, it doesn't really matter much, but using the first `res` instead of `BUF_SIZE` for the second `splice()` might be a bit clearer to someone who hasn't seen this syscall before.  This change is definitely up to you though.

Reply:
In that case, the very last call to the second `splice()` could return less than the value returned by the first `splice()`.  Then, because there’s no more data to read, the first `splice()` will return 0 and the loop will end, meaning the pipe won’t be fully flushed and thus the byte count will be wrong.

If you use the value returned by the first `splice()`, everything should work out fine.  Also, that’s the value we actually care about, since we want to see how many bytes the input file contains rather than how many get successfully copied to `/dev/null`.

None of this matters if the second `splice()` will always return the same value as the first `splice()` though.